### PR TITLE
Poseidon implementation

### LINKED ISSUE #22: Implement Poseidon test vectors

### PR DESCRIPTION
**Pull request by therealyingtong**
Closes #22.
 
Uses test vectors from https://github.com/daira/pasta-hadeshash/commit/cdab5252d0ec675780e9f8c3d09fdb7a0a71d3ff.

### CODE REVIEW

**Review by daira — APPROVED** [review]
ACK. Feel free to add more test vectors!

File: orchard_poseidon.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+# Number of partial rounds
+R_P = 58
+# Width
+t = 3
+
+# Initial capacity element
+CAPACITY_ELEMENT = Fp(1 << 65)
+
+def poseidon_hash(x, y):
+    return perm([x, y, CAPACITY_ELEMENT])
```
Comment (str4d) [inline orchard_poseidon.py:15]: Move these into `orchard_poseidon_hash.py` and then add test vector generation (e.g. some sequential inputs and some random) both here for `perm` and in `orchard_poseidon_hash.py` for `poseidon_hash`.

File: orchard_poseidon_hash.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    # Generate random test vectors
+    for _ in range(10):
+        test_vectors.append([
+            Fp(leos2ip(rand.b(32))),
+            Fp(leos2ip(rand.b(32))),
+        ])
+
+    render_tv(
+        render_args(),
+        'poseidon_hash',
```
Comment (str4d) [inline orchard_poseidon_hash.py:36]: This is used in a comment; it should be the filename.
```suggestion
        'orchard_poseidon_hash',
```

File: orchard_poseidon_hash.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+from tv_output import render_args, render_tv
+from tv_rand import Rand
+
+# Initial capacity element
+CAPACITY_ELEMENT = Fp(1 << 65)
+
+def poseidon_hash(x, y):
+    assert isinstance(x, Fp)
+    assert isinstance(y, Fp)
+    return perm([x, y, CAPACITY_ELEMENT])
```
Comment (str4d) [inline orchard_poseidon_hash.py:13]: The hash function only returns the first element of the state:
```suggestion
    return perm([x, y, CAPACITY_ELEMENT])[0]
```

File: orchard_poseidon_hash.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            Fp(leos2ip(rand.b(32))),
+            Fp(leos2ip(rand.b(32))),
+        ])
+
+    render_tv(
+        render_args(),
+        'poseidon_hash',
+        (
+            ('input', '[[u8; 32]; 2]'),
+            ('output', '[[u8; 32]; 3]'),
```
Comment (str4d) [inline orchard_poseidon_hash.py:39]:
```suggestion
            ('output', '[u8; 32]'),
```

File: orchard_poseidon_hash.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        ),
+        [{
+            'input': list(map(bytes, input)),
+            'output': list(map(bytes, poseidon_hash(input[0], input[1]))),
```
Comment (str4d) [inline orchard_poseidon_hash.py:43]:
```suggestion
            'output': bytes(poseidon_hash(input[0], input[1])),
```

File: orchard_poseidon.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    for _ in range(10):
+        test_vectors.append([
+            Fp(leos2ip(rand.b(32))),
+            Fp(leos2ip(rand.b(32))),
+            Fp(leos2ip(rand.b(32)))
+        ])
+
+    render_tv(
+        render_args(),
+        'poseidon_perm',
```
Comment (str4d) [inline orchard_poseidon.py:316]:
```suggestion
        'orchard_poseidon',
```

File: orchard_poseidon.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+R_P = 58
+# Width
+t = 3
+
+def perm(input_words):
+    R_f = int(R_F / 2)
+
+    round_constants_counter = 0
+
+    state_words = list(input_words)
```
Comment (str4d) [inline orchard_poseidon.py:19]:
```suggestion
    state_words = list(input_words)
    assert len(state_words) == t
```

**Review by str4d — CHANGES_REQUESTED** [review]

File: orchard_poseidon.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
+from orchard_pallas import Fp
+import numpy as np
```
Comment (str4d) [inline orchard_poseidon.py:2]: Document this dependency in `README.md`.

File: orchard_poseidon.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+    Fp(0x0b01e425270ec4f3d4810b0395cd422ac96ed1a3c8eedc8c8b02bc51ec68a656),
+    Fp(0x133f84bf2658613979ceca9514f5c4c3fe8448e7d92088c17ebcffa2b473a674),
+    Fp(0x3cbdf49b42e4c2a512987260f2327ea02739e1e2db9a462a7fe1f95e8b8c3dd2),
+    Fp(0x2d38b041fc5379cc67db3544a2bf6a17d63396b16af9a2ba30492c0982ebed07),
+    Fp(0x3ae6480f450154a60be0da12998d3bdeadfcfee0c8fac2167c091bfb3d26dd25),
+    Fp(0x01ace40f25423f2c2a7005641a6a1d2d3066e72bb10b457010579b58eac52105),
+    Fp(0x1212fa565936b14ffea2b74ba21379ccb28ce22d6ad899932ef97c7899c68028),
+    Fp(0x2166389a0e4e1f364627f44f183c91e19a0f302972d3b542faae23893d5f50fd),
+    Fp(0x0ade42a791fef5cf2c386d2df993e84538576cf565d379e8a3be16b562d755c2)
+]
```
Comment (str4d) [inline orchard_poseidon.py:256]:
```suggestion
]
assert len(ROUND_CONSTANTS) == (R_F + R_P) * t
```

File: orchard_poseidon.py [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+            state_words[i] = (state_words[i]).exp(5)
+        state_words = list(np.array(MDS_MATRIX).dot(np.array(state_words, dtype=object)))
+    
+    return state_words
+
+# Round constants generated by the reference implementation script, commit 659de89
+# https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/generate_parameters_grain.sage
+#
+# $ sage generate_parameters_grain.sage 1 0 255 3 8 58 0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001
+ROUND_CONSTANTS = [
```
Comment (str4d) [inline orchard_poseidon.py:57]: I checked this.

File: orchard_poseidon.py [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+    Fp(0x1212fa565936b14ffea2b74ba21379ccb28ce22d6ad899932ef97c7899c68028),
+    Fp(0x2166389a0e4e1f364627f44f183c91e19a0f302972d3b542faae23893d5f50fd),
+    Fp(0x0ade42a791fef5cf2c386d2df993e84538576cf565d379e8a3be16b562d755c2)
+]
+
+# MDS matrix generated by the reference implementation script, commit 659de89
+# https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/generate_parameters_grain.sage
+#
+# $ sage generate_parameters_grain.sage 1 0 255 3 8 58 0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001
+MDS_MATRIX = [
```
Comment (str4d) [inline orchard_poseidon.py:262]: I checked this.

File: orchard_poseidon.py [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+    render_tv(
+        render_args(),
+        'orchard_poseidon',
+        (
+            ('input', '[[u8; 32]; 3]'),
+            ('output', '[[u8; 32]; 3]'),
+        ),
+        [{
+            'input': list(map(bytes, input)),
+            'output': list(map(bytes, perm(input))),
```
Comment (str4d) [inline orchard_poseidon.py:324]:
```suggestion
            ('initial_state', '[[u8; 32]; 3]'),
            ('final_state', '[[u8; 32]; 3]'),
        ),
        [{
            'initial_state': list(map(bytes, input)),
            'final_state': list(map(bytes, perm(input))),
```

**Review by str4d — APPROVED** [review]
ACK after comments are addressed.

**Review by daira — COMMENTED** [review]
Post-hoc utACK.

**Pull request merged**